[Lollo] Step2 - 음료수 3계층 생성 & 자판기 구조체 생성 & XCTest 시도

eeeesong opened the pull request:
### 작업 목록

- [x] class init 방식 개선 (enum 사용하지 않기, 외부에서 값 받아오기)
- [x] 음료수 class 재구성 - 3단계 계층 만들기
- [x] 음료수 하위 타입의 특성에 따른 protocol 생성, 채택
- [x] VendingMachine struct 구현
- [x] VendingMachine의 기능에 따라 객체 단위로 분화 (Beverages / Organizer / MoneyBox)
- [ ] 객체 생성 시 팩토리 메소드 사용
- [x] XCTest로 VendingMachine 기능 테스트 (Mock 활용)



### 학습 키워드

🔑 **Protocol**: `Default Implement`, `associatedtype` , `Hashable`

🔑 **인스턴스 생성**: `Factory Method Pattern` , `Generic`

🔑 **XCTest**: `XCTest`, `Unit Test`

🔑 **Date**: `Compare Date` 



### 고민과 해결

- 음료수 객체 하위 타입에 **프로토콜** 채택 시 중복되는 메소드 **구현부 중복**을 줄일 수 없을 지 고민해보았습니다.

  - `Default Implement`를 알게되어 프로토콜 extension을 이용해보려고 시도했습니다. 
  - 그러나 해당 메소드가 생성 된 타입 내의 프로퍼티를 메소드 내부에서 사용하려다 보니 애매해져서 적용하지 못했습니다...

- **프로토콜** 메소드에 필요한 **프로퍼티** 구현에 대해서도 고민해보았습니다.

  - 수업 때 `Associated Type`에 대해 말씀해주신 것을 참고하여 해당 타입을 선언해주었는데, 결국은 프로토콜에서 public으로 변수를 선언할 수밖에 없게 되는 듯하여 사용하지 않았습니다.

- **VendingMachine struct 설계**에 대해 고민했습니다.

  - 처음에는 필요한 프로퍼티, 메소드를 전부 VendingMachine에 작성했는데 양도 많고 다양한 분류의 것들이 한 데에 있는 것이 좋지 않다고 판단하여 역할에 따라 객체를 나누어 보았습니다.
    - Beverages: 창고 역할을 하는 타입으로, 재고와 관련한 모든 것을 처리하도록 했습니다.
    - Organizer: 특정 기준에 따라 제품을 분류하는 기능을 모두 모았습니다.
    - MoneyBox: 돈과 관련한 역할을 모아두었습니다.

- **VendingMachine의 재고 저장** 방식에 대해서도 고민을 해보았습니다.

  - 재고를 저장하는 데이터 타입으로는 dictionary가 가장 적당해보였고, 그러기 위해선 key 값이 필요했습니다.
  - 따라서 Beverage에 Hashable 프로토콜을 채택하도록 하고, 해시 값을 제품 이름으로 설정했습니다.
  - XCTest를 진행하다보니 dictionary 타입이 순서 보장이 되지 않아서 문제가 발생한다는 것을 확인했습니다.
  - Comparable 프로토콜을 추가적으로 채택하고, 이름에 따라 순서를 정렬할 수 있도록 하였습니다. 

- **팩토리 메소드 패턴** 시도

  - 지난 PR 때 말씀해주신 팩토리 메소드 패턴을 알아본 뒤, 적용하기 위해 아래와 같은 프로토콜의 생성을 시도해보았습니다

    - Shopable: 자판기에서 팔 수 있는 제품의 추상화 (Beverage에 대응)
    - Storagable: 자판기에 저장할 수 있는 저장소의 추상화 (Beverages에 대응)

  - `associatedType Item: Shopable & Hashable` 을 통해 **Storagable** 프로토콜에 타입을 지정해준 뒤 필요한 메소드에 해당 타입을 사용했습니다. 그리고 **StorageFactory** 프로토콜을 생성하여 Storagable 프로토콜을 따르는 인스턴스 생성 메소드를 만드려고 했는데, 아래와 같은 오류에 막혀서 끝까지 진행시키지 못했습니다...

    ```swift
    Protocol 'Storagable' can only be used as a generic constraint because it has Self or associated type requirements
    ```

  - 구글링을 해봐도 잘 이해가 되지 않았는데, generic에 대한 이해가 부족하여 그런 것 같아서 차차 더 공부를 한 뒤 재시도 해보려고 합니다..

  - 🥲 고민만 많이 하고 성공한 게 없는 느낌이네요 하하..



### 질문거리

- 코드에 주석으로도 남겨 놓았는데요, MoneyBox에서 잔액을 가져오는 것의 경우 `func balance() -> Int { return money }` 와 같은 방식을 벗어나지 못했습니다. 해당 값으로 뭔가 다른 일을 하는 것이 아니라서 다른 방법을 생각하지 못했는데, 혹시 이런 경우에는 뾰족한 수가 있을까요..?

eeeesong committed:
README: Step 1 구현 내용 업데이트

eeeesong committed:
Date 포맷 변경 메소드 개선

eeeesong committed:
Beverage 하위 계층의 특성에 따른 protocol 선언

eeeesong committed:
Milk, Soda, Coffee class 개선 / 프로토콜 채택, 구현

- Milk, Soda, Coffee: 각 클래스에 중첩으로 존재하던 Kind enum 삭제, 외부에서 값을 넘겨 초기화하도록 변경. 각 타입에 맞는 프로토콜 채택 후 기능 구현
- DateGenerator: 유통기한을 구하기 위한 메소드 추가. (특정 날짜로 부터 n일 이후 Date 리턴)

eeeesong committed:
Milk, Soda, Coffee의 subclass 구현

- 각 특성에 따라 프로토콜을 채택하도록 함

eeeesong committed:
Beverage 관련 프로토콜 / 데이터타입을 별개의 파일로 분리

eeeesong committed:
VendingMachine struct 구현

- 관리자 기능: 재고 관리, 물품 관리 등
- 사용자 기능: 돈, 구매 관련
- 음료수  필터: 특정 특성을 만족하는 음료수 반환

eeeesong committed:
음료수 상위 클래스 기능 추가

- 가격 관련 메소드 추가
- Hashable 프로토콜 채택

eeeesong committed:
Beverages class 생성: 재고 관리 기능 위임

eeeesong committed:
Organizer / MoneyBox 생성: 관련 기능 위임

- Organizer: 특정 기준에 따라 상품을 거르는 필터 기능 관련 집합
- MoneyBox: 돈 관련 기능 집합

eeeesong committed:
VendingMachine 기능 분화에 따른 수정

eeeesong committed:
Date 관련 기능 정리, 파일명 변경

eeeesong committed:
XCTest: 유닛 테스트 생성

eeeesong committed:
Beverage를 이름으로 정렬할 수 있도록 변경 (Comparable)

eeeesong committed:
VendingMachine 테스트 시나리오 실행

eeeesong committed:
사소한 수정: print 구문 삭제, 메소드 이름 변경

godrm commented on VendingMachineApp/VendingMachineApp/Beverage.swift:
```diff
@@ -22,8 +22,32 @@ class Beverage: CustomStringConvertible {
+
+extension Beverage: Comparable {
+    static func < (lhs: Beverage, rhs: Beverage) -> Bool {
+        return lhs.name > rhs.name
```
부등호가 반대인데 이게 맞게 구현된거죠?

eeeesong replied:
아 아뇨.. 잘못 썼습니다. XCTest 작성하다가 추가해서 정신이 없었나봅니다🥲

godrm commented on VendingMachineApp/VendingMachineApp/Beverage.swift:
```diff
@@ -22,8 +22,32 @@ class Beverage: CustomStringConvertible {
+}
+
+extension Beverage: Hashable {
+    static func == (lhs: Beverage, rhs: Beverage) -> Bool {
+        return lhs.name == rhs.name
+    }
+    
+    func hash(into hasher: inout Hasher) {
+      hasher.combine(name)
+    }
```
hash(into:) 는 어떤 경우에 구현해야 하는건가요?

eeeesong replied:
Dictionary Key로 지정하기 위한 해시값을 구하는 용도입니다!

godrm replied:
네. 그럴 때 정말 name만 비교하면 되는지 고려해보세요. 지금은 동작할텐데 항상 고민해봐야 하는 포인트입니다.

godrm commented on VendingMachineApp/VendingMachineApp/Beverage.swift:
```diff
@@ -22,8 +22,32 @@ class Beverage: CustomStringConvertible {
+    
+    func bought(_ moneyUpdator: (Int) -> ()) {
+        return moneyUpdator(-price)
+    }
```
이런 방식 괜찮네요. 다만 -price가 좀 어색하긴 합니다 ㅎㅎ 빼준다는 것은 알겠는데... 저도 좀 고민해보겠습니다.

eeeesong replied:
`let value = price * -1` 이런식으로 따로 값을 정의해주면 좀 나을까요??

godrm replied:
클로저와 값을 빼는 부분이 어색해보였어요. 그냥 매개변수로 잔액을 넘기고 price를 빼서 남은 금액을 리턴하는 것도 괜찮겠네요.

eeeesong replied:
매개변수로 잔액 넘기는 방향으로 변경했습니다

godrm commented on VendingMachineApp/VendingMachineApp/BeverageProtocols.swift:
```diff
@@ -0,0 +1,38 @@
+import Foundation
+
+
+protocol Transportable {
+    
+    func isTransportable() -> Bool
+    
+}
+
+protocol Expirable {
+    
+    func isExpired() -> Bool
+    
+}
+
+protocol LactoFreeable {
+    
+    func isLactoFree() -> Bool
+}
+
+protocol SugarFreeable {
+    
+    func isSugarFree() -> Bool
+    
+}
+
+protocol Hotable {
+    
+    func isHotable() -> Bool
+    
+}
```
이게 다 Beverage 음료에 적용하는 프로토콜인가요? 
별도로 선언해도 되고 Beverage에 선언해도 됩니다. 참고하세요. 어느 타입과 관련이 있는지 아직은 모르겠어요

godrm commented on VendingMachineApp/VendingMachineApp/Beverages.swift:
```diff
@@ -0,0 +1,68 @@
+        checkSoldout(for: beverage)
+    }
+    
+    private func checkSoldout(for beverage: Beverage) {
```
soldout 여부를 판단한다면 check-를 붙이지 않아도 되지 않을까요? 군더더기 표현같아서요 
그리고 판단하는 게 아닌가요? return 타입이 없어도 되는건가요?

eeeesong replied:
음 메소드 이름을 잘못 지은 것 같습니다
재고가 0이 되는 경우를 판단해서 특정 dictionary key를 nil로 바꾸어 준다 -> 상황을 판단해서 sold out 팻말을 붙인다
이런 식으로 생각을 했던 건데 좀 더 좋은 네이밍을 생각해보겠습니다

eeeesong replied:
updateSoldout으로 변경하였습니다

godrm commented on VendingMachineApp/VendingMachineApp/Beverages.swift:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+    
+    func pullOut(_ beverage: Beverage) -> Beverage? {
```
이 부분은 조금 어색한데요, 이미 해당 음료타입 인스턴스가 있는데 동일한 인스턴스를 찾는건가요?
타입에 대한 메타정보를 활용하는 방법도 있습니다. 참고하세요.

eeeesong replied:
`타입에 대한 메타정보` 이 부분 이해를 잘 못했습니다. 참조 비교 같은 걸 이용하라는 말씀이실까요..?
매개변수와 반환값으로 모두 Beverage를 준 이유는 구매하려는 음료가 인자로 전달된 후, 그 전달 값에 해당하는 음료가 있으면 반환하여
재고목록에서 팔린목록으로 옮기기 위함이었습니다!

godrm replied:
클래스와 인스턴스 관계를 생각해보세요. Beverage 클래스와 실제로 만들어진 인스턴스 관계를 보면 인스턴스들은 다 같은 Beverage 타입이겠죠. (실제로는 세부로 나뉜 서브 클래스겠죠) 
인스턴스 메타 정보는 클래스가 됩니다. 클래스의 메타 정보는 뭘까요? 그걸 메타 클래스라고 부릅니다 ㅎㅎ 
예를 들어 10 과 8 은 다른 값이지만 같은 Int 타입이죠. 이걸 비교하려면 어떻게 해야할까요? 같은 클래스인지 메타 정보를 비교해야 합니다. 
Beverage에 대해서도 그런게 가능할까요?

eeeesong replied:
오 이제 무슨 말씀인지 알겠습니다. 
다만 어떻게 적용해야할 지는 좀 더 생각을 해봐야겠네요😅
자세히 설명해주셔서 감사합니다!!

eeeesong replied:
이 부분은 의미있는 변경 방향이 떠오르지 않아서 변경하지 못했습니다ㅠㅠ

godrm commented on VendingMachineApp/VendingMachineApp/Beverages.swift:
```diff
@@ -0,0 +1,68 @@
+            return filter(stockList)
+        } else {
+            return storage
+        }
```
참고로 클로저도 default type이 가능합니다. 
전체를 전달하는 것과 필터링하는 것과 구분할 방법은 없을까요? 
모든 메소드를 클로저 방식으로 구현해야 하는 건 아닙니다. 그냥 전체 아이템을 리턴해주는 메소드가 나을 수도 있죠.

eeeesong replied:
전체 전달과 필터링 메소드를 구분하였습니다

godrm commented on VendingMachineApp/VendingMachineApp/Coffee.swift:
```diff
@@ -7,31 +7,76 @@
+        case .bottle:
+            return true
+        default:
+            return false
```
isTransportable() 같은 메소드가 있는 것 좋네요. 프로토콜로 선언한 것도 좋습니다. 
- isHot 같은 경우는 Bool 타입보다는 온도를 속성으로 갖고, 기준 온도보다 낮은지 높은지 판단해주는 메소드가 있는게 좋습니다. 
- 속성과 메소드를 설계할 때는 속성이 모든 것을 판단하기보다는 메소드가 그 값을 처리하는 방법을 같이 고민해보세요.

eeeesong replied:
속성 관련 메소드를 전반적으로 수정하였습니다
매개변수를 활용하여 값을 처리하도록 변경하였습니다

godrm commented on VendingMachineApp/VendingMachineApp/Coffee.swift:
```diff
@@ -7,31 +7,76 @@
+    }
+    
+    func isSugarFree() -> Bool {
+        return !isSugarAdded
```
여기도 isSugarAdded를 설탕 함유량 같은 데이터를 저장하고 판단하는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Coffee.swift:
```diff
@@ -7,31 +7,76 @@
+    }
+    
+    func isExpired() -> Bool {
+        let today = Date()
+        let isExpired = expiry < today
```
이런 값도 매개변수로 넘길 수 있도록 하는게 더 유연한 구조가 됩니다.

godrm commented on VendingMachineApp/VendingMachineApp/Coffee.swift:
```diff
@@ -7,31 +7,76 @@
+    }
+    
+    func isLactoFree() -> Bool {
+        return lactose <= 0.5 ? true : false
```
이런 비교하는 메소드도 매개변수와 기준 값을 상수로 선언하는 것을 고려하세요.

godrm commented on VendingMachineApp/VendingMachineApp/DateExtensions.swift:
```diff
@@ -0,0 +1,35 @@
+        let date = formatter.date(from: string)
+        
+        return date ?? Date()
+    }
```
이런 메소드는 Date를 생성하는 거니까, 팩토리 메소드와 비슷하네요.
팩토리 메소드 패턴을 찾아서 학습해보세요.
Date 타입을 생성하는 데 인스턴스 메소드 형태가 좋은지도 살펴보세요.

eeeesong replied:
DateFactory를 생성하고 각 class의 초기화 방식을 업데이트하였습니다

godrm commented on VendingMachineApp/VendingMachineApp/DateExtensions.swift:
```diff
@@ -0,0 +1,35 @@
+        let expiryDate = Calendar.current.date(byAdding: .day,
+                                               value: amount,
+                                               to: self)
+        return expiryDate ?? self
```
amount 단위가 날인지 시간인지 함수만 보고 판단하기 어렵습니다. 그 부분을 명확하게 표현해주면 좋겠습니다.

eeeesong replied:
명확하게 보이도록 days로 변경하였습니다

godrm commented on VendingMachineApp/VendingMachineApp/Milk.swift:
```diff
@@ -8,31 +8,71 @@
-                   price: info.price,
-                   size: info.size,
-                   manufactured: info.manufactured)
+class Milk: Beverage, Expirable, LactoFreeable {
```
프로토콜로 확장하는 방식도 매우 좋습니다만
서브 클래스를 만들면서 들어가는 속성이 있으면 같이 만들어지는 메소드들은 서브 클래스에서 처리해도 됩니다. 
만약 프로토콜을 Milk 이외 다른 타입에서도 사용한다면 충분히 의미가 있겠네요. 좋은 접근입니다.

godrm commented on VendingMachineApp/VendingMachineApp/MoneyBox.swift:
```diff
@@ -0,0 +1,26 @@
+
+class MoneyBox {
+    
+    private var money: Int
```
돈 자체를 추상화하는 타입이 있다는 시도는 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachine.swift:
```diff
@@ -7,21 +7,70 @@
     
-    private var beverages: [Beverage]
+    private var inList: Beverages
+    private var outList: Beverages
```
outList는 구매한 것만 보관하는 건가요? inList와 outList가 한 번에 딱 떠오르지는 않네요.

eeeesong replied:
네 맞습니다 inList는 자판기 창고에 있는 목록, outList는 팔린 목록을 의미한 거였는데 네이밍을 고민해봐야겠네요!

eeeesong replied:
이 부분도 productList, soldList로 이름 변경하였습니다

godrm commented on VendingMachineApp/VendingMachineApp/Organizer.swift:
```diff
@@ -0,0 +1,100 @@
+
+import Foundation
+
+class Organizer {
```
Organizer 이게 VendingMachine과 어떤 기준으로 나눴는지 궁금합니다. 
이상하다기 보다는 하위 타입을 선언하면 자판기가 아니라 다른 자동화 기계에서도 쓸 수 있을정도로 추상화가 된 것인가 싶어서 질문을 남겨봅니다. 
계산하는 부분에 대한 책임만 나눈걸까요? 아니면 음료를 관리하는 로직만 나눈걸까요? 
Organizer 이름은 조금 일반적인 명사라서 세부적인 동작을 유추하기는 어려워서요. 
그리고 아래 메소드들을 보면 moneyBox 라는 속성과 관련있는 동작보다 자판기에 있는 음료 목록을 받아서 처리하는 메소드가 많아서 이 타입만의 동작이라고 보기 어렵네요.

eeeesong replied:
Organizer는 특정 필터에 따라서 목록을 반환하는 일을 하는 걸로만 묶었습니다. 
쇼핑몰로 치자면 `가격 순으로 정렬` / `인기 순으로 정렬` / `가격대에 따라 정렬` 이런 역할만 하는 친구로 생각을 하고 설계를 했습니다
Organizer로 쓰기 전엔 Curator로 생각을 했었는데요, 오히려 그게 나을까요??

godrm replied:
이름 자체보다도 속성과 메소드의 관계를 잘 살펴보면 좋겠습니다. 응집도가 없어보여서요.

eeeesong replied:
속성과 메소드의 관계를 살펴보니 moneyBox에는 돈 관련한 메소드가 모이는 게 좋을 것 같아서 구매 가능한 목록을 리턴하는 메소드도 해당 클래스로 이동하였습니다.
그리고 Organizer의 이름을 Filterer로 바꾸고, 설탕 기준치 등 구분의 기준이 되는 정보들을 프로퍼티로 가지고 있도록 변경하였습니다.

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachine.swift:
```diff
@@ -7,21 +7,70 @@
+    
+    func healthyItems() -> [Beverage] {
+        return organizer.healthyItems(from: inList)
+    }
```
이렇게 구분해놓은 것은 매우 깔끔하고 좋습니다.

godrm reviewed (changes requested):
수고하셨습니다. 다음 피드백 내용을 확인해보세요.

eeeesong committed:
Beverage 특성 관련 Protocol 메소드의 기준 판단 방식 변경

eeeesong committed:
VendingMachine 포함 class 내용 수정

- Filterer: Organizor에서 이름 변경, 관련 프로토콜 메소드 변경에 따른 메소드 변경, 관련 프로퍼티 추가
- MoneyBox: 돈 관련한 메소드 응집하도록 수정

eeeesong committed:
사소한 수정: 프로퍼티/ 메소드 이름 변경 등

eeeesong committed:
DateFactory 생성 / Date 관련 수정

- 관련 class 이니셜라이저 수정
- Date add 메소드의 매개변수 이름 변경 (amount -> days)

eeeesong committed:
기존 테스트 내용 삭제

eeeesong commented:
코멘트 주신 부분들 수정하였습니다.
pullOut에 대해 코멘트 주신 부분은 해결책을 찾지 못하여 반영하지 못했습니다..
그리고 유닛테스트는 오늘 수업에서 말씀하신 부분들을 아직 반영하지 못해서 우선 지웠는데요,
혹시 이번 단계에서 의존성을 약하게 만드는 변경을 모두 하고 넘어가는 것이 좋을까요??

godrm commented on VendingMachineApp/VendingMachineApp/DateFactory.swift:
```diff
@@ -0,0 +1,38 @@
+    
+}
+
+class SixDigitFactory: DateFactory {
```
아.. dateFactory가 뭔가 했는데 Date 생성하는 팩토리 객체였네요.
Date 팩토리는 Date 타입이 하는 것과 어떤 차이가 있을까요? 
새로운 시도였는데 고민해볼 포인트 같습니다.

godrm commented:
수고하셨습니다. 
의존성에 관련해서는 꾸준히 단계별로 작업해보세요.
